docs/pages/pmd/devdocs/adding_metrics_support_to_language.md

```diff
@@ -0,0 +1,125 @@
+than that. For instance, the project mirror is very language specific. Java's implementation uses the natural structure
+provided by the language's package system to structure the project's content. Apex, on the other, has no package 
+system and thus can't use the same mechanism. That explains why the interfaces framing the project mirror are very 
+loose. Their main goal is to provide type safety through generics.
+ 
+Moreover, usage resolution depends on the availability of type resolution for the given language, which is only implemented in 
+Java. For these reasons, signature matching is considered an optional feature of the metrics framework. But despite 
+this limitation, signature matching still provides a elegant way to find information about the class we're in. This 
+feature requires no usage resolution and can be used to implement sophisticated metrics, that already give access to 
+detection strategies.
+
+## Implementation of a new framework
+
+### 1. Groundwork
+
+* Create a class implementing `QualifiedName`. This implementation must be tailored to the target language so 
+that it can indentify unambiguously any class and operation in the analysed project (see `JavaQualifiedName`). You 
+must implement `equals`, `hashCode` and `toString`.
+* Determine the AST nodes that correspond to class and method declaration in your language. These types are 
+referred hereafter as `T` and `O`, respectively. Both these types must implement the interface `QualifiableNode`, which 
+means they must expose a `getQualifiedName` method to give access to their qualified name.
+
+### 2. Implement the project memoizer
+* Create a class extending `BasicProjectMemoizer<T, O>`. That's all.
+
+### 3. Implement the façade
+* Create a class extending `AbstractMetricsComputer<T, O>`. This object will be responsible for calculating metrics 
+given a memoizer, a node and info about the metric. Typically, this object is stateless so you might as well make it 
+a singleton.
+* Create a class extending `AbstractMetricsFacade<T, O>`. This class needs a reference to your `ProjectMemoizer` and 
+your `MetricsComputer`. It backs the real end user façade, and handles user provided parameters before delegating to 
+your `MetricsComputer`.
+* Create the static façade of your framework. This one has an instance of your `MetricsFaçade` object and delegates 
+static methods to that instance.
+* Create classes `AbstractOperationMetric` and `AbstractClassMetric`. These must implement `Metric<T>` and 
+`Metric<O>`, respectively. They typically provide defaults for the `supports` method of each metric. 
+* Create enums `ClassMetricKey` and `OperationMetricKey`. These must implement `MetricKey<T>` and `MetricKey<O>`. The
+ enums list all available metric keys for your language.
+* Create metrics by extending your base classes, reference them in your enums, and you can start using them with your
+ façade!
+
```

Comment:
I think, we need a 4th step here, to be complete: "Wiring the metric framework".
This is just the step, to implement `getMetricsVisitorFacade` in the specific language module.